jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java

```diff
@@ -0,0 +1,540 @@
+        CorrelationKey correlationKey = correlationKeyFactory.newCorrelationKey(caseId);
+        
+        Collection<org.jbpm.services.api.model.NodeInstanceDesc> nodes = runtimeDataService.getNodeInstancesByCorrelationKeyNodeType(correlationKey, 
+                                                                                                Arrays.asList(ProcessInstance.STATE_ACTIVE), 
+                                                                                                Arrays.asList("MilestoneNode"), 
+                                                                                                queryContext);
+        
+        Collection<Long> completedNodes = nodes.stream().filter(n -> ((NodeInstanceDesc)n).getType() == 1).map(n -> n.getId()).collect(toList());
+        Predicate<org.jbpm.services.api.model.NodeInstanceDesc> filterNodes = null;
+        if (achievedOnly) {            
+            filterNodes = n -> ((NodeInstanceDesc)n).getType() == 1;             
+        } else {
+            filterNodes = n -> ((NodeInstanceDesc)n).getType() == 0;
+        }
+        Collection<CaseMilestoneInstance> milestones = nodes.stream()
+        .filter(filterNodes)
+        .map(n -> new CaseMilestoneInstanceImpl(String.valueOf(n.getId()), n.getName(), completedNodes.contains(n.getId()), n.getDataTimeStamp()))
+        .collect(toList());
+        
+        return milestones;
+    }
+
+    @Override
+    public Collection<CaseStageInstance> getCaseInstanceStages(String caseId, boolean activeOnly, QueryContext queryContext) {
+        ProcessInstanceDesc pi = runtimeDataService.getProcessInstanceByCorrelationKey(correlationKeyFactory.newCorrelationKey(caseId));        
+        if (pi == null || !pi.getState().equals(ProcessInstance.STATE_ACTIVE)) {
+            throw new CaseNotFoundException("No case instance found with id " + caseId + " or it's not active anymore");
+        }
+        
+        CaseDefinition caseDef = getCase(pi.getDeploymentId(), pi.getProcessId());
+        Collection<CaseStageInstance> stages = internalGetCaseStages(caseDef, caseId, activeOnly, queryContext);
+        
+        return stages;
+    }
+    
+    /*
+     * Case instance queries
+     */
+    
+
+    @Override
+    public Collection<org.jbpm.services.api.model.NodeInstanceDesc> getActiveNodesForCase(String caseId, QueryContext queryContext) {
+        Map<String, Object> params = new HashMap<String, Object>();
+        params.put("caseId", caseId);
+        applyQueryContext(params, queryContext);
+        applyDeploymentFilter(params);
+        List<org.jbpm.services.api.model.NodeInstanceDesc> nodeInstances =  commandService.execute(new QueryNameCommand<List<org.jbpm.services.api.model.NodeInstanceDesc>>("getActiveNodesForCase", params));
+        return nodeInstances;
+    }
+
+
+    @Override
+    public Collection<CaseInstance> getCaseInstances(QueryContext queryContext) {
```

Comment:
The method below could be reused passing the default statuses, avoiding code dup.